contrib/rabbitmq/pkg/reconciler/rabbitmqsource.go

```diff
@@ -0,0 +1,172 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	controllerAgentName = "rabbitmq-source-controller"
+	raImageEnvVar       = "RABBITMQ_RA_IMAGE"
+	finalizerName       = controllerAgentName
+)
+
+func Add(mgr manager.Manager, logger *zap.SugaredLogger) error {
+	raImage, defined := os.LookupEnv(raImageEnvVar)
+	if !defined {
+		return fmt.Errorf("required environment variable '%s' not defined", raImageEnvVar)
+	}
+
+	log.Println("Adding the Rabbitmq Source controller.")
+	p := &sdk.Provider{
+		AgentName:  controllerAgentName,
+		Parent:     &v1alpha1.RabbitmqSource{},
+		Owns:       []runtime.Object{&v1.Deployment{}},
+		Reconciler: &reconciler{
+			scheme:              mgr.GetScheme(),
+			receiveAdapterImage: raImage,
+		},
+	}
+
+	return p.Add(mgr, logger)
+}
+
+type reconciler struct {
+	client              client.Client
+	scheme              *runtime.Scheme
+	receiveAdapterImage string
+}
+
+func (r *reconciler) InjectClient(c client.Client) error {
+	r.client = c
+	return nil
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, object runtime.Object) error {
+	logger := logging.FromContext(ctx).Desugar()
+
+	src, ok := object.(*v1alpha1.RabbitmqSource)
+	if !ok {
+		logger.Error("could not find Rabbitmq source", zap.Any("object", object))
+		return nil
+	}
+
+	logger.Info("Reconciling new source: ", zap.Any("source", src))
+
+	src.Status.InitializeConditions()
+
+	sinkURI, err := sinks.GetSinkURI(ctx, r.client, src.Spec.Sink, src.Namespace)
+	if err != nil {
+		src.Status.MarkNoSink("NotFound", "")
+		return err
+	}
+
+	src.Status.MarkSink(sinkURI)
+
+	_, err = r.createReceiveAdapter(ctx, src, sinkURI)
+	if err != nil {
+		logger.Error("Unable to create the receive adapter", zap.Error(err))
+		return err
+	}
+	src.Status.MarkDeployed()
+
+	return nil
+}
+
+func (r *reconciler) createReceiveAdapter(ctx context.Context, src *v1alpha1.RabbitmqSource, sinkURI string) (*v1.Deployment, error) {
+	ra, err := r.getReceiveAdapter(ctx, src)
+	if err != nil && !apierrors.IsNotFound(err) {
+		logging.FromContext(ctx).Error("Unable to get an existing receive adapter", zap.Error(err))
+		return nil, err
+	}
+	if ra != nil {
+		logging.FromContext(ctx).Desugar().Info("Reusing existing receive adapter", zap.Any("receiveAdapter", ra))
+		return ra, nil
+	}
+	svc := resources.MakeReceiveAdapter(&resources.ReceiveAdapterArgs{
+		Image:   r.receiveAdapterImage,
+		Source:  src,
+		Labels:  getLabels(src),
+		SinkURI: sinkURI,
+	})
+
+	if err := controllerutil.SetControllerReference(src, svc, r.scheme); err != nil {
+		return nil, err
+	}
+
+	err = r.client.Create(ctx, svc)
+	logging.FromContext(ctx).Desugar().Info("Receive Adapter created.", zap.Error(err), zap.Any("receiveAdapter", svc))
+	return svc, err
+}
+
+func (r *reconciler) getReceiveAdapter(ctx context.Context, src *v1alpha1.RabbitmqSource) (*v1.Deployment, error) {
+	dl := &v1.DeploymentList{}
```

Comment:
My _recommendation_ (not a requirement), is that you use a fixed name + Get, rather than GenerateName + list. I know that most (all?) the existing examples use the GenerateName + list pattern, but we are going to remove that as it is somewhat flaky (see knative/eventing#1250).